File: app/models/automation.py
from beanie import Document, Indexed
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class CampaignTarget(str, Enum):
    USERS = "users"
    VOLUNTEERS = "volunteers"
    BOTH = "both"
    CUSTOM = "custom"  # Custom email list


class CampaignStatus(str, Enum):
    DRAFT = "draft"
    GENERATING = "generating"  # AI is generating email
    READY = "ready"  # Email generated, ready to send
    SENDING = "sending"  # Currently sending
    SENT = "sent"  # All emails sent
    FAILED = "failed"


class EmailRecipient(BaseModel):
    email: str
    name: str
    user_type: str  # "user", "volunteer"
    sent: bool = False
    sent_at: Optional[datetime] = None
    error: Optional[str] = None


class EmailCampaign(Document):
    """Email automation campaign"""

    # Campaign details
    campaign_name: str
    organization_id: str
    created_by_user_id: str

    # Target audience
    target_audience: CampaignTarget
    event_id: Optional[str] = None  # If targeting specific event volunteers

    # Email content
    email_subject: Optional[str] = None
    email_context: Optional[str] = None  # User's brief/context for AI
    generated_email_html: Optional[str] = None  # AI-generated HTML email
    generated_email_plain: Optional[str] = None  # Plain text version

    # Recipients
    total_recipients: int = 0
    recipients: List[EmailRecipient] = []

    # Sending progress
    emails_sent: int = 0
    emails_failed: int = 0

    # Status
    status: CampaignStatus = CampaignStatus.DRAFT

    # Logs (for real-time progress)
    logs: List[str] = []

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    sent_at: Optional[datetime] = None

    class Settings:
        name = "email_campaigns"
        indexes = [
            "organization_id",
            "created_by_user_id",
            "status",
            "target_audience",
            "created_at",
        ]

    def add_log(self, message: str):
        """Add a log entry with timestamp"""
        timestamp = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
        self.logs.append(f"[{timestamp}] {message}")

    class Config:
        json_schema_extra = {
            "example": {
                "campaign_name": "Monthly Newsletter",
                "target_audience": "users",
                "email_context": "Send update about new events this month",
                "status": "draft"
            }
        }

File: src/services/automation.js
import api from './api'

// Step 1: Create Campaign & Extract Recipients
export const createCampaign = async (campaignName, targetAudience, emailContext, eventId = null) => {
  const response = await api.post('/automation/campaigns', {
    campaign_name: campaignName,
    target_audience: targetAudience, // "users" | "volunteers" | "both"
    event_id: eventId,
    email_context: emailContext
  })
  return response.data
}

// Step 2: Generate Email with Groq AI
export const generateEmail = async (campaignId, emailSubject) => {
  const response = await api.post(`/automation/campaigns/${campaignId}/generate-email`, {
    email_subject: emailSubject
  })
  return response.data
}

// Step 3: Preview Email
export const previewEmail = async (campaignId) => {
  const response = await api.get(`/automation/campaigns/${campaignId}/preview`)
  return response.data
}

// Step 4: Send Campaign
export const sendCampaign = async (campaignId) => {
  const response = await api.post(`/automation/campaigns/${campaignId}/send`)
  return response.data
}

// Get all campaigns
export const getCampaigns = async () => {
  const response = await api.get('/automation/campaigns')
  return response.data
}

// Get campaign details
export const getCampaignDetails = async (campaignId) => {
  const response = await api.get(`/automation/campaigns/${campaignId}`)
  return response.data
}

// Delete campaign
export const deleteCampaign = async (campaignId) => {
  const response = await api.delete(`/automation/campaigns/${campaignId}`)
  return response.data
}
